accounts/accounts.go

```diff
@@ -46,41 +47,67 @@ func NewAccount(username string, hostname string) *Account {
 		Directory: directory,
 		Passwd:    "generated password",
 		NextId:    1,
+		Token:     token,
 	}
 }
 
-// TODO: Not threadsafe, or persistent
-func (account *Account) IncrementId() {
-	// TODO: Needs to be done correctly
-	account.NextId += 1
-}
+var current *Authentication
 
 func NewAuthentication() *Authentication {
+	if current != nil {
+		return current
+	}
+
 	account_map := &AccountMap{
 		Accounts: make(map[string]*Account, 0),
 	}
 
-	return &Authentication{
+	current = &Authentication{
 		Map: account_map,
 		db:  datastore.NewDatabase("Accounts"),
 	}
+
+	current.load()
+
+	return current
+}
+
+func FindAccount(username string) *Account {
+	auth := NewAuthentication()
+	account := auth.Find(username)
+	return account
 }
 
 // Update the datastructure in memory
 func (auth *Authentication) UpdateAccount(account *Account) {
 
 	auth.mux.Lock()
+	auth.file_lock()
 
 	fmt.Println("Updating Account")
+
+	auth.load() // Could have changed since last loaded
 	auth.Map.Accounts[account.Username] = account
+	auth.store()
 
+	auth.file_unlock()
 	auth.mux.Unlock()
 }
 
-func (auth *Authentication) Find(username string) *Account {
+func (auth *Authentication) Reload() {
```

Comment:
This whole Authentication struct looks pretty complicated and error prone with multiple locks, reloads, etc. Why is it necessary? Is it possible to simplify the authentication layer to avoid using files, etc? Like mentioned before, it is okay (and even preferable for the sake of the challenge) to keep the authentication layer as simple as possible and accept some pre-generated credentials on the CLI.